### PR TITLE
Invocation interception

### PR DESCRIPTION
**Pull request by mathewc**
Alternate proposal to https://github.com/Azure/azure-webjobs-sdk/pull/1167. In this case, a handler can be specified as part of the trigger data.

### CODE REVIEW

File: src/Microsoft.Azure.WebJobs.Host/Executors/FunctionInstanceFactory.cs [hunk 1]
```diff
@@ -21,10 +20,10 @@ public FunctionInstanceFactory(IFunctionBinding binding, IFunctionInvoker invoke
             _descriptor = descriptor;
         }
 
-        public IFunctionInstance Create(Guid id, Guid? parentId, ExecutionReason reason, IDictionary<string, object> parameters)
+        public IFunctionInstance Create(FunctionInstanceFactoryContext context)
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Executors/FunctionInstanceFactory.cs]: The main change is just an internal refactoring to use a context object here, which is a good change to make anyways

File: src/Microsoft.Azure.WebJobs.Host/Executors/TriggeredFunctionData.cs [hunk 2]
```diff
@@ -19,5 +20,10 @@ public class TriggeredFunctionData
         /// The trigger value for a specific triggered function invocation.
         /// </summary>
         public object TriggerValue { get; set; }
+
+        /// <summary>
+        /// Optional handler function for processing the invocation.
+        /// </summary>
+        public Func<Func<Task>, Task> InvokeHandler { get; set; }
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Executors/TriggeredFunctionData.cs]: This is the key - the listener when it calls the executor can specify this handler

**Comment by cgillum** [comment]
From the trigger author's perspective, it seems this would be very easy to use.  It solves my problem, so I'm all for it. :)

File: src/Microsoft.Azure.WebJobs.Host/Executors/IFunctionInvoker.cs [hunk 3]
```diff
@@ -1,16 +1,23 @@
...
 using System.Threading.Tasks;
 
-namespace Microsoft.Azure.WebJobs.Host.Executors
+namespace Microsoft.Azure.WebJobs.Host
 {
-    internal interface IFunctionInvoker
+    /// <summary>
+    /// Provides an interface for invoking a job function.
+    /// </summary>
+    public interface IFunctionInvoker
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Executors/IFunctionInvoker.cs]: Making this public. We'll also be able to replace https://github.com/Azure/azure-webjobs-sdk-script/blob/1ab54a94916c8180105af1e43b24a483f070a53f/src/WebJobs.Script/Description/IFunctionInvoker.cs#L9 with this.

File: src/Microsoft.Azure.WebJobs.Host/Executors/IFunctionInvokerProvider.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace Microsoft.Azure.WebJobs.Host
+{
+    /// <summary>
+    /// Defines an interface for the creation of <see cref="IFunctionInvoker"/> instances.
+    /// </summary>
+    public interface IFunctionInvokerProvider
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Executors/IFunctionInvokerProvider.cs]: A custom binding can implement this interface to provide custom invokers.

**Pull request merged**